crates/wal-protocol/src/lib.rs

```diff
@@ -225,3 +265,274 @@ impl MatchKeyQuery for Envelope {
+            codec: StorageCodecKind,
+            value: &T,
+        ) -> Result<Self, StorageEncodeError> {
+            let mut buf = BytesMut::new();
+            storage::encode::encode_serde(value, &mut buf, codec)?;
+
+            Ok(Self {
+                codec: Some(codec),
+                bytes: buf.freeze(),
+            })
+        }
+
+        fn encode_bilrost<T: bilrost::Message>(value: &T) -> Result<Self, StorageEncodeError> {
+            Ok(Self {
+                codec: Some(StorageCodecKind::Bilrost),
+                bytes: storage::encode::encode_bilrost(value),
+            })
+        }
+
+        fn encode_protobuf<T: prost::Message>(value: &T) -> Result<Self, StorageEncodeError> {
+            let mut buf = BytesMut::new();
+            value
+                .encode(&mut buf)
+                .map_err(|err| StorageEncodeError::EncodeValue(err.into()))?;
+
+            Ok(Self {
+                codec: Some(StorageCodecKind::Protobuf),
+                bytes: buf.freeze(),
+            })
+        }
+
+        fn decode_serde<T: DeserializeOwned>(mut self) -> Result<T, StorageDecodeError> {
+            let codec = self.codec()?;
+            if !matches!(
+                codec,
+                StorageCodecKind::Json
+                    | StorageCodecKind::FlexbuffersSerde
+                    | StorageCodecKind::BincodeSerde
+            ) {
+                return Err(StorageDecodeError::UnsupportedCodecKind(codec));
+            }
+
+            storage::decode::decode_serde(
+                &mut self.bytes,
+                self.codec.ok_or(DecodeError::MissingFieldCodec)?,
+            )
+        }
+
+        fn decode_bilrosst<T: bilrost::OwnedMessage>(mut self) -> Result<T, StorageDecodeError> {
+            let codec = self.codec()?;
+            if codec != StorageCodecKind::Bilrost {
+                return Err(StorageDecodeError::UnsupportedCodecKind(codec));
+            }
+
+            storage::decode::decode_bilrost(&mut self.bytes)
+        }
+
+        fn decode_protobuf<T: prost::Message + Default>(self) -> Result<T, StorageDecodeError> {
+            let codec = self.codec()?;
+            if codec != StorageCodecKind::Protobuf {
+                return Err(StorageDecodeError::UnsupportedCodecKind(codec));
+            }
+
+            T::decode(self.bytes).map_err(|err| StorageDecodeError::DecodeValue(err.into()))
+        }
+
+        fn codec(&self) -> Result<StorageCodecKind, DecodeError> {
+            self.codec.ok_or(DecodeError::MissingFieldCodec)
+        }
+    }
+
+    #[derive(bilrost::Message)]
+    struct Envelope {
+        #[bilrost(1)]
+        header: Field,
+        #[bilrost(2)]
+        command_kind: CommandKind,
+        #[bilrost(3)]
+        command: Field,
+    }
+
+    pub fn encode(envelope: &super::Envelope) -> Result<Bytes, StorageEncodeError> {
+        // todo(azmy): avoid clone? this will require change to `From` implementation
```

Comment:
I'm a little concerned about those extra clones.

Reply:
Same here, this is why exactly I want to make sure there is no performance regression. I will report the numbers after running the bench